File: python/read.py
from google.protobuf.internal.encoder import _VarintBytes
from google.protobuf.internal.decoder import _DecodeVarint32

import metric_pb2
import random

with open('out.bin', 'rb') as f:
    buf = f.read()
    n = 0
    count = 0
    l = 0

    #while n < len(buf):
    ptr = buf
    while l < len(buf):
        msg_len, new_pos = _DecodeVarint32(ptr, 0)
        n = new_pos
        msg_buf = ptr[n:n+msg_len]
        n += msg_len
        l += n
        ptr = ptr[n:]
        read_metric = metric_pb2.Metric()
        read_metric.ParseFromString(msg_buf)
        print(read_metric)
        count += 1

    print(count)

File: go/read.go
package main

import (
	"github.com/golang/protobuf/proto"
	"encoding/binary"
	"errors"
    "io"
    "os"
    "log"
    pb "grpc_serialize_example/proto"
)

var errInvalidVarint = errors.New("invalid varint32 encountered")

func ReadDelimited(r io.Reader, m proto.Message) (n int, err error) {
	// Per AbstractParser#parsePartialDelimitedFrom with
	// CodedInputStream#readRawVarint32.
	var headerBuf [binary.MaxVarintLen32]byte
	var bytesRead, varIntBytes int
	var messageLength uint64
	for varIntBytes == 0 { // i.e. no varint has been decoded yet.
		if bytesRead >= len(headerBuf) {
			return bytesRead, errInvalidVarint
		}
		// We have to read byte by byte here to avoid reading more bytes
		// than required. Each read byte is appended to what we have
		// read before.
		newBytesRead, err := r.Read(headerBuf[bytesRead : bytesRead+1])
		if newBytesRead == 0 {
			if err != nil {
				return bytesRead, err
			}
			// A Reader should not return (0, nil), but if it does,
			// it should be treated as no-op (according to the
			// Reader contract). So let's go on...
			continue
		}
		bytesRead += newBytesRead
		// Now present everything read so far to the varint decoder and
		// see if a varint can be decoded already.
		messageLength, varIntBytes = proto.DecodeVarint(headerBuf[:bytesRead])
	}
	messageBuf := make([]byte, messageLength)
	newBytesRead, err := io.ReadFull(r, messageBuf)
	bytesRead += newBytesRead
	if err != nil {
		return bytesRead, err
	}
	return bytesRead, proto.Unmarshal(messageBuf, m)
}

func main() {
	in, err := os.Open("out.bin")
    if err != nil {
        log.Fatalln("Error reading file:", err)
    }

    for {
        metric := &pb.Metric{}
        n, err := ReadDelimited(in, metric)
        if n == 1 {
            break
        }
        if err != nil {
            log.Printf("err %s", err)
        }
        log.Printf("metric: %d %s %s %s", n, metric.Name, metric.Type, metric.Tags)
    }
}
